### ISSUE TITLE
Asst: Sequencing not displaying the correct sanger wells

### ISSUE THREAD

**Issue by santiagochrist** [bthuronyi/CloneCoordinate#264]
![Screenshot 2024-10-30 at 11 19 52 PM](https://github.com/user-attachments/assets/38b1bbfc-5e38-405e-81a9-a3bbb0dd3919)

**Comment by bthuronyi** [comment]
Looks like the text got cut off which is the main issue. You've got duplicate sort numbers here, which it can't deal with. We want the error about that to be clearly visible though, no matter how many duplicate numbers there are. So it looks like that code does need updating!

**Comment by santiagochrist** [comment]
This is actually an issue with the main tab and not the assistant, so I removed that label.

So Column `AB` seems to pull from the hidden Column `BK`, so I've gone ahead and edited that column and added a few new lines. Currently, the text wrap is set to overflow and it can span columns `X:AB`, so the error and the numbers should be visible.

I changed this line: `"Error!"&char(10)&"repeated well ID in unsubmitted samples! "&char(10)&join(",",sort(serialnos))` from the full column code below.
```
=iferror(
let(serialnos,iferror(filter(BK3:BK,AC3:AC<>"",AG3:AG="",regexmatch(E3:E,"Sanger"),not(regexmatch(H3:H,"[Oo]mit"))),#VALUE!),
ifs(
iserror(serialnos),1,
counta(serialnos)=0,1,
join(",",unique(sort(serialnos)))<>join(",",sort(serialnos)),"Error!"&char(10)&"repeated well ID in unsubmitted samples! "&char(10)&join(",",sort(serialnos)),
max(serialnos)=counta(serialnos),counta(serialnos)+1,
TRUE,xmatch(FALSE,arrayformula(sequence(counta(serialnos))=sort(serialnos)))
)),1)
```
Output
-
<img width="587" alt="Screenshot 2025-01-01 at 11 35 49 AM" src="https://github.com/user-attachments/assets/ce0de2b0-0479-43f5-ba2a-f01b0d7b79dd" />

**Comment by santiagochrist** [comment]
I modified the message to add a redundant "!" so we can remove that if we want to keep the format consistent. Maybe we can substitute the second "!" with a ":"

Anyway, let me know what you think @bthuronyi

**Comment by bthuronyi** [comment]
I think the problem remains if a large number of wells are repeated. We need to truncate the output AND focus on just the repeated wells. Here's what you see if a dozen or so wells are repeated between unsubmitted samples (done by deleting a bunch of "date submitted" entries:
<img width="519" alt="Screenshot 2025-01-12 at 9 39 29 AM" src="https://github.com/user-attachments/assets/9fe4f5fc-6b28-4f13-88cf-25afd15a18e4" />

**Comment by bthuronyi** [comment]
The error message text gets lost and no amount of overflow will help if the number of samples can be arbitrarily large. So let's instead show only, let's say, the first 3 repeated wells (or serial numbers) alongside the error.

**Comment by bthuronyi** [comment]
It could also be simpler to just return the first repeated serial number and leave it at that. Then as users try to fix it, they'll either have to whack-a-mole or will themselves find the pattern and sort it out. That's not perfect but it's not bad.

**Comment by santiagochrist** [comment]
This took me way too long but turns out `SORTN` is the function to go with here.

Column BK Code
-
This line is doing the magic here: `join(",",sortn(unique(serialnos),3))`
```
=iferror(
let(serialnos,iferror(filter(BK3:BK,AC3:AC<>"",AG3:AG="",regexmatch(E3:E,"Sanger"),not(regexmatch(H3:H,"[Oo]mit"))),#VALUE!),
ifs(
iserror(serialnos),1,
counta(serialnos)=0,1,
join(",",unique(sort(serialnos)))<>join(",",sort(serialnos)),"Error!"&char(10)&"repeated well ID in unsubmitted samples! "&char(10)&join(",",sortn(unique(serialnos),3)),
max(serialnos)=counta(serialnos),counta(serialnos)+1,
TRUE,xmatch(FALSE,arrayformula(sequence(counta(serialnos))=sort(serialnos)))
)),1)

**Comment by santiagochrist** [comment]
Before: 
![image](https://github.com/user-attachments/assets/9ee58311-557d-4992-a367-9249b873d95f)


After:
![image](https://github.com/user-attachments/assets/5d0bc17a-2053-4e1b-aee3-a25ebc3b01f9)

**Comment by bthuronyi** [comment]
That's awesome, I never heard of SORTN before, seems really useful! Whenever I've had to do that, I've always done it in super kludgy awful ways instead. I feel like this could be really useful, especially if you force it NOT to sort but instead to just give the first n elements of the array...

Anyway: that tackles the "don't overload the number of repeat well IDs" problem, but we would still want in this case to display ONLY the repeated elements, or at least the first few of them. That's more helpful for addressing the problem than giving the unique well numbers.

**Comment by bthuronyi** [comment]
Try this:
`=let(uniques,unique(A:A,false,true),unique({uniques;sequence(max(uniques))},false,true))`

where A:A is the sort numbers range.

UNIQUE(data,false,true) sorts by row and then only outputs the values that are unique -- removing any that are duplicated. So 1 2 2 3 → 1 3.

Then we can add to that array sequence(max(uniques)) which provides all the numbers in order, and UNIQUE(data,false,true) it AGAIN to give only the numbers that WERE duplicated in the original input.

Not the most elegant or efficient but it looks like it works.

**Comment by santiagochrist** [comment]
We are going to need to write this one down...`array_constrain`. Try out this algorithm:

```=join(",",array_constrain(unique(filter(A:A, countif(A:A, A:A) > 1)),3,1))```

**Comment by santiagochrist** [comment]
`array_constrain` cuts down an array to x rows and y columns, but it can effect be parsed into a comma-separated list consisting of the first n characters.

**Comment by santiagochrist** [comment]
So right now I have: `join(",",array_constrain(unique(filter(serialnos, countif(serialnos, serialnos) > 1)),3,1))`

![image](https://github.com/user-attachments/assets/8be76de2-3470-4d70-919a-3c45b0cdd057)

**Comment by bthuronyi** [comment]
> `array_constrain` cuts down an array to x rows and y columns, but it can effect be parsed into a comma-separated list consisting of the first n characters.

Nice -- this I had seen before but forgot about... always learning!

**Comment by bthuronyi** [comment]
> unique(filter(A:A, countif(A:A, A:A) > 1))

Interesting... is this sneakily subbing one of the COUNTIF references to A:A with each element in A:A in turn (b/c of FILTER) while the other one is looking at the entire A:A array? I guess it must be that way so that it checks the count of each element of A:A in turn.

This looks good, please implement in main CC.

**Comment by santiagochrist** [comment]
Yep, that's more or less what I was trying to get at. Works in CC fantastically!
